packages/medusa/src/api/store/products/helpers.ts

```diff
@@ -30,3 +47,101 @@ export const maybeApplyStockLocationId = async (req: MedusaRequest, ctx) => {
 
   return entities.map((entity) => entity.stock_location_id)
 }
+
+export const wrapProductsWithTaxPrices = async <T>(
```

Comment:
We discussed the middleware approach briefly with @carlos-r-l-rodrigues yesterday evening, and he had a few good thoughts on this. Sharing here this morning, so we can kick off the discussion before his AM:

- Performance for store product list and retrieve endpoints is critical, so we should be careful about introducing too many operations at the HTTP layer that would harm it. Tax calculation is a potentially heavy operation. 
- Performing updates on prices at the HTTP layer will make the sorting capabilities in the search module (when introduced) less useful. You can't rely on the sorting from the search engine because you are modifying your prices at a higher level in the stack.

I think both are valid points and good to open a discussion around.

Right now (without the search module), I can see at least two paths to go down:
- Leave the middleware as is, knowing we will have to rethink (and refactor) it when introducing the search module
- Centralise product variant price queries to a single endpoint, e.g. `GET /store/products/prices?id[]=...`. This would mean a product page and product grid would to perform two requests to get the commonly required information to build such pages (as suggested by @carlos-r-l-rodrigues)
```
GET /store/products
GET /store/products/prices?id[]=...
```

Maybe there are other approaches to consider. Feel free to pitch in.

I don't have any strong opinion on this, there are pros and cons to both approaches. However, I do see the benefit of having a separate endpoint for price calculations in case we want to introduce more dynamic stuff in that context. Another case to consider in the future is inventory which will need to be computed and decorated dynamically on every request. Depending on what we decide here, we should try to follow a similar approach.

-----------------------

**On the search module**

We can index anything we want into the search module, but in the first iteration, it relies on the entity module events (correct me if I am wrong @adrien2p, @carlos-r-l-rodrigues). Further down the line, we could imagine that the engine would expose "pre-indexing hooks", allowing us to perform calculations on resources before they are stored in the engine DB. 

An example use case of such pre-indexing hook would be price calculations. We could calculate product variant prices for each region and currency, so that queries like the one we are discussing would require only a single request to the search engine.

Reply:
@olivermrbl I agree on having the products endpoint very performant. In this case, unless you ask for the calculation by passing a cart or country context (we can even be more explicit), no calculations will happen, so it won't have an impact unless you need that information. 

Adding a separate endpoint won't solve anything, as it will have the exact same problems as doing `fields=*variants.calculated_price`. Doing that will actually introduce more issues
1. It will make the interface more complicated. 
2. The FE will need to do a join between products and prices
3. Sorting by price is a very common action, and that won't work on the `store/products` endpoint anymore.

If we want this tax calculation, we either do it like this, or we need to move that calculation to the pricing module. However, that won't reduce the number of operations, as you still need to list the tax lines and pass them to the pricing module.

Basically, our architecture doesn't allow for significant performance improvements at the core. The main performance gains will have to come from a good caching layer.

Reply:
If the resources are indexed in the search module, the request should be (still to be tested) highly performant and support sorting across modules, e.g. sort product variant prices. If we can index the prices with taxes taken into account, I think that would be ideal and probably what we should aim for to solve this. The two approaches I outlined above are both temporary imo., I just want us to consider which one we feel most comfortable with in the near term. There is no pretty way to do this kind of stuff right now, which is why we have all of this middleware acrobatics going on. 

I agree we will need to introduce heavy usage of caching, but I think that's a separate topic in this context and something we will need to do irregardless of this.

Reply:
I think the idea with the separate endpoint was also to have a place to introduce even more dynamic calculations in the future, in case we need that. E.g. dynamic to the point of being real-time, which wouldn't be possible to index in the search engine.

Reply:
Can we transform this into a workflow to fetch products?
This does not seem to be the case of middlewares, this is a cross domain calculation. It is more business logic than anything else at this point.

Reply:
I know where you are coming from, and I agree with having most of the BL outside the HTTP layer, but I don't think putting it in a workflow is the right solution here (or for any read-only operations). What we could do is create a simple helper function that wraps all of this logic and put it higher up the stack, WDYT?